### PR TITLE
Implement SPV_INTEL_token_type extension

### PR DESCRIPTION
**Pull request by MrSidims**
This extension adds a type-declaration instruction OpTypeTokenINTEL,
which is an analog of LLVM's type token. This type specifies IDs of
entry points and exit points of certain IL constructs, for example it
can be useful for exception handling representation in IL.

Spec: https://github.com/intel/llvm/pull/3788

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
@AlexeySotkin @AlexeySachkov  @Fznamznon please take a look

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll [hunk 1]
```diff
@@ -0,0 +1,37 @@
+;; Check whether the translator reports an error for a module with token type
+;; if SPV_INTEL_token_type extension is not used.
```
Comment (Fznamznon) [inline test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll:2]: Is second semicolon required?
```suggestion
; Check whether the translator reports an error for a module with token type
; if SPV_INTEL_token_type extension is not used.
```

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+; RUN: llvm-as < %s -o %t.bc
+; RUN: not llvm-spirv %t.bc 2>&1 --spirv-allow-unknown-intrinsics | FileCheck %s
+
+; CHECK: RequiresExtension: Feature requires following SPIR-V extension:
```
Comment (Fznamznon) [inline test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll:7]: I think it is also good to say which exact feature requires extension. i.e. something like this:
```
RequiresExtension: Feature:
declare token @llvm.tokenfoo()
requires following SPIR-V extension:
SPV_INTEL_token_type

File: lib/SPIRV/SPIRVWriter.cpp [hunk 3]
```diff
@@ -293,6 +293,13 @@ SPIRVType *LLVMToSPIRVBase::transType(Type *T) {
   if (T->isFloatingPointTy())
     return mapType(T, BM->addFloatType(T->getPrimitiveSizeInBits()));
 
+  if (T->isTokenTy()) {
```
Comment (Fznamznon) [inline lib/SPIRV/SPIRVWriter.cpp:296]: Just curious, what used to happen when token type was used in input module?

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll (hunk 2 above)
Comment (MrSidims) [reply test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll:7]: Kind of applied, guess it's the best, what I can do here

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll (hunk 1 above)
Comment (MrSidims) [reply test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll]: Nope, it's just my habit (may be bad) to but double semicolon to add a comment in LIT. Applied.

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll (hunk 2 above)
Comment (Fznamznon) [reply test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll:7]: That is good enough for me, thanks!

**Review by Fznamznon — APPROVED** [review]

File: lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -470,7 +470,7 @@ Type *SPIRVToLLVM::transType(SPIRVType *T, bool IsClassMember) {
 
   SPIRVDBG(spvdbgs() << "[transType] " << *T << " -> ";)
   T->validate();
-  switch (T->getOpCode()) {
+  switch (static_cast<size_t>(T->getOpCode())) {
```
Comment (AlexeySotkin) [inline lib/SPIRV/SPIRVReader.cpp:473]: Minor: Why `size_t`? AFAIK on most systems `size_t` has 64-bit width, but 32 bits should be enough for OpCode.

File: test/transcoding/SPV_INTEL_token_type/token_type_intel.ll [hunk 5]
```diff
@@ -0,0 +1,44 @@
+; RUN: llvm-as < %s -o %t.bc
+; RUN: llvm-spirv %t.bc -o %t.spv --spirv-allow-unknown-intrinsics --spirv-ext=+SPV_INTEL_token_type
+; RUN: llvm-spirv %t.spv -to-text -o - | FileCheck %s --check-prefix=CHECK-SPIRV
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis %t.rev.bc -o - | FileCheck %s --check-prefix=CHECK-LLVM
+
+; CHECK-SPIRV: Capability TokenTypeINTEL
+; CHECK-SPIRV: Extension "SPV_INTEL_token_type"
+; CHECK-SPIRV: Name [[FUN:[0-9]+]] "llvm.tokenfoo"
```
Comment (AlexeySotkin) [inline test/transcoding/SPV_INTEL_token_type/token_type_intel.ll:9]: Nit: I think [Numeric Substitution Blocks](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-numeric-substitution-blocks) look more readable:
```suggestion
; CHECK-SPIRV: Name [[#FUN:]] "llvm.tokenfoo"
```

File: test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll [hunk 6]
```diff
@@ -0,0 +1,39 @@
+; Check whether the translator reports an error for a module with token type
```
Comment (AlexeySotkin) [inline test/transcoding/SPV_INTEL_token_type/token_type_intel_negative.ll:1]: Please move the test to `test/negative/` directory

File: lib/SPIRV/libSPIRV/SPIRVErrorEnum.h [hunk 7]
```diff
@@ -17,3 +17,4 @@ _SPIRV_OP(InvalidInstruction, "Can't translate llvm instruction:\n")
 _SPIRV_OP(InvalidWordCount,
           "Can't encode instruction with word count greater than 65535:\n")
 _SPIRV_OP(Requires1_1, "Feature requires SPIR-V 1.1 or greater:")
+_SPIRV_OP(RequiresExtension, "Feature requires following SPIR-V extension:\n")
```
Comment (AlexeySotkin) [inline lib/SPIRV/libSPIRV/SPIRVErrorEnum.h:20]:
```suggestion
_SPIRV_OP(RequiresExtension, "Feature requires the following SPIR-V extension:\n")
```

**Comment by MrSidims** [comment]
Applied the suggestions. @AlexeySotkin please ping me if you want me to merge the commits together.

File: test/negative/SPV_INTEL_token_type/token_type_intel_negative.ll [hunk 8]
```diff
@@ -0,0 +1,39 @@
+; Check whether the translator reports an error for a module with token type
```
Comment (AlexeySotkin) [inline test/negative/SPV_INTEL_token_type/token_type_intel_negative.ll:1]: Does it make sense to have SPV_INTEL_token_type directory with only one file? Should we have just `test/negative/token_type.ll`?

File: test/negative/feature_requires_extension.ll [hunk 9]
```diff
@@ -0,0 +1,39 @@
+; Check whether the translator reports an error for a module with token type
```
Comment (MrSidims) [inline test/negative/feature_requires_extension.ll:1]: Changed the test name, but left this comment on purpose as an example what the test is checking

**Review by AlexeySotkin — APPROVED** [review]

**Comment by MrSidims** [comment]
@AlexeySotkin done

**Review by AlexeySotkin — APPROVED** [review]

**Pull request merged**